mopidy/config/types.py

```diff
@@ -178,30 +219,116 @@ def serialize(self, value, display=False):
+    Values can optionally not be a pair, in which case the whole input is provided for
+    both sides of the value.
+    """
+
+    def __init__(
+        self, optional=False, optional_pair=False, separator="|", subtypes=None
+    ):
+        self._required = not optional
+        self._optional_pair = optional_pair
+        self._separator = separator
+        if subtypes:
+            self._subtypes = subtypes
+        else:
+            self._subtypes = (String(), String())
+
+    def deserialize(self, value):
+        raw_value = decode(value).strip()
+        validators.validate_required(raw_value, self._required)
+        if not raw_value:
+            return None
+
+        if self._separator in raw_value:
+            values = raw_value.split(self._separator, 1)
+        elif self._optional_pair:
+            values = (raw_value, raw_value)
+        else:
+            raise ValueError(
+                f"Config value must include {self._separator!r} separator: {raw_value}"
+            )
+
+        return (
+            self._subtypes[0].deserialize(encode(values[0])),
+            self._subtypes[1].deserialize(encode(values[1])),
+        )
+
+    def serialize(self, value, display=False):
+        serialized_first_value = self._subtypes[0].serialize(
+            value[0], display=display
+        )
+        serialized_second_value = self._subtypes[1].serialize(
+            value[1], display=display
+        )
+
+        if (
+            not display
+            and self._optional_pair
+            and serialized_first_value == serialized_second_value
+        ):
+            return serialized_first_value
+        else:
+            return "{0}{1}{2}".format(
+                serialized_first_value,
+                self._separator,
+                serialized_second_value,
+            )
+
+
 class List(ConfigValue):
     """List value.
 
-    Supports elements split by commas or newlines. Newlines take presedence and
+    Supports elements split by commas or newlines. Newlines take precedence and
     empty list items will be filtered out.
     """
 
-    def __init__(self, optional=False):
+    def __init__(self, optional=False, unique=False, subtype=None):
         self._required = not optional
+        self._unique = unique
+        self._subtype = subtype if subtype else String()
 
     def deserialize(self, value):
         value = decode(value)
         if "\n" in value:
             values = re.split(r"\s*\n\s*", value)
         else:
             values = re.split(r"\s*,\s*", value)
-        values = tuple(v.strip() for v in values if v.strip())
+
+        # This is necessary for backwards-compatibility, in case subclasses
+        # aren't calling their parent constructor.
+        subtype = getattr(self, "_subtype", String())
+
+        values_iter = (
+            subtype.deserialize(v.strip()) for v in values if v.strip()
+        )
+        if self._unique:
+            values = frozenset(values_iter)
```

Comment:
This may return a different order compared to the value. I think that would be unexpected.

Reply:
:thinking:

Reply:
How much of a concern is it likely to be? The behaviour is opt-in, so I guess we can always adjust the behaviour later based on feedback if absolutely necessary.

Reply:
If it was documented it'd be better. I just don't like the surprise.

Reply:
I've added a comment :+1: